internal/cephfs/core/metadata.go

```diff
@@ -0,0 +1,177 @@
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unknown ||
+		clusterAdditionalInfo[s.clusterID].subVolMetadataState == supported {
+		fsa, err := s.conn.GetFSAdmin()
+		if err != nil {
+			return "", err
+		}
+
+		res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
+		if err == nil {
+			clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
+
+			return res, nil
+		}
+		var invalid fsAdmin.NotImplementedError
+		// In case the error is other than invalid command return error to the caller.
+		if !errors.As(err, &invalid) {
+			return "", err
+		}
+	}
+	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
+
+	return "", fmt.Errorf("operation GetMetadata is not supported")
+}
+
+// SetMetadata sets custom metadata on the subvolume in a volume as a
+// key-value pair.
+func (s *subVolumeClient) SetMetadata(key, value string) error {
+	// keyPresent checks whether corresponding clusterID key is present in clusterAdditionalInfo
+	var keyPresent bool
+	// verify if corresponding clusterID key is present in the map,
+	// and if not, initialize with default values(false).
+	if _, keyPresent = clusterAdditionalInfo[s.clusterID]; !keyPresent {
+		clusterAdditionalInfo[s.clusterID] = &localClusterState{}
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unknown ||
+		clusterAdditionalInfo[s.clusterID].subVolMetadataState == supported {
+		fsa, err := s.conn.GetFSAdmin()
+		if err != nil {
+			return err
+		}
+
+		err = fsa.SetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key, value)
+		if err == nil {
+			clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
+
+			return nil
+		}
+		var invalid fsAdmin.NotImplementedError
+		// In case the error is other than invalid command return error to the caller.
+		if !errors.As(err, &invalid) {
+			return err
+		}
+	}
+	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
+
+	return fmt.Errorf("operation SetMetadata is not supported")
+}
+
+// RemoveMetadata removes custom metadata set on the subvolume in a volume
+// using the metadata key.
+func (s *subVolumeClient) RemoveMetadata(key string) error {
+	// keyPresent checks whether corresponding clusterID key is present in clusterAdditionalInfo
+	var keyPresent bool
+	// verify if corresponding clusterID key is present in the map,
+	// and if not, initialize with default values(false).
+	if _, keyPresent = clusterAdditionalInfo[s.clusterID]; !keyPresent {
+		clusterAdditionalInfo[s.clusterID] = &localClusterState{}
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unknown ||
```

Comment:
Move the logic of checking support to a helper function and reuse it?

Reply:
No, your previous suggestion for code change is good.